builder/static_code.go

```diff
@@ -888,41 +951,208 @@ func listJoin(list []string, sep string, lastSep string) string {
 	}
 }
 
-func (p *parser) parseRule(rule *rule) (any, bool) {
+// ==template== {{ if .LeftRecursion }}
+
+func (p *parser) checkPrevChoice(rule *rule) (bool, bool) {
+	var currentEStack []any
+	var prevEStack []any
+	for i := 1; i <= len(p.rstack); i++ {
+		indexCurrent := len(p.rstack) - i
+		indexPrev := len(p.rstack) - i*2
+		if indexPrev < 0 {
+			continue
+		}
+		if p.rstack[indexCurrent].rule == rule && p.rstack[indexPrev].rule == rule {
+			currentEStack = p.rstack[indexCurrent].estack
+			prevEStack = p.rstack[indexPrev].estack
+			break
+		}
+	}
+	if prevEStack == nil || currentEStack == nil {
+		return false, false
+	}
+	if len(prevEStack) != len(currentEStack) {
+		panic("Stacks are not equal(len)")
+	}
+
+	for i := len(prevEStack) - 1; i >= 0; i-- {
+		currentCh, ok := currentEStack[i].(*choiceExpr)
+		if !ok {
+			continue
+		}
+		prevCh, ok := prevEStack[i].(*choiceExpr)
+		if !ok {
+			panic("Stacks are not equal(position choiceExpr)")
+		}
+		if currentCh != prevCh {
+			panic("Stacks are not equal(choiceExpr)")
+		}
+		currentAlt := -1
+		for j, inExp := range currentCh.alternatives {
+			if inExp == currentEStack[i+1] {
+				currentAlt = j
+				break
+			}
+		}
+		if currentAlt == -1 {
+			panic("lost alternatives in choiceExpr")
+		}
+		prevAlt := -1
+		for j, inExp := range prevCh.alternatives {
+			if inExp == prevEStack[i+1] {
+				prevAlt = j
+				break
+			}
+		}
+		if prevAlt == -1 {
+			panic("lost alternatives in choiceExpr")
```

Comment:
When does this happen? I am unsure about the usage of panic here.

Reply:
This should never happen. This is a check for the correct state of the parser. The chosen alternative that is on the stack must be present in choiceExpr. Indicates that an error has been made in the code.